network/p2pMetainfo.go

```diff
@@ -0,0 +1,126 @@
+// due to msgp allocbound enforcement we need to limit the number of headers and values
+// but it is cannot be done without msgp modification to accept both map and slice allocbound
+// so that introduce a service peerMetaValues type to have allocbound set and msgp generator satisfied.
+
+const maxHeaderKeys = 64
+const maxHeaderValues = 16
+
+// SortString is a type that implements sort.Interface for sorting strings
+type SortString = basics.SortString
+
+//msgp:allocbound peerMetaValues maxHeaderValues
+type peerMetaValues []string
+
+//msgp:allocbound peerMetaHeaders maxHeaderKeys
+type peerMetaHeaders map[string]peerMetaValues
+
+func peerMetaHeadersToHTTPHeaders(headers peerMetaHeaders) http.Header {
+	httpHeaders := make(http.Header, len(headers))
+	for k, v := range headers {
+		httpHeaders[k] = v
+	}
+	return httpHeaders
+}
+
+func peerMetaHeadersFromHTTPHeaders(headers http.Header) peerMetaHeaders {
+	pmh := make(peerMetaHeaders, len(headers))
+	for k, v := range headers {
+		pmh[k] = v
+	}
+	return pmh
+}
+
+type peerMetaInfo struct {
+	telemetryID  string
+	instanceName string
+	version      string
+	features     string
+}
+
+func readPeerMetaHeaders(stream io.ReadWriter, p2pPeer peer.ID, netProtoSupportedVersions []string) (peerMetaInfo, error) {
+	var msgLenBytes [2]byte
+	rn, err := stream.Read(msgLenBytes[:])
+	if rn != 2 || err != nil {
+		err0 := fmt.Errorf("error reading response message length from peer %s: %w", p2pPeer, err)
+		return peerMetaInfo{}, err0
+	}
+
+	msgLen := binary.BigEndian.Uint16(msgLenBytes[:])
+	msgBytes := make([]byte, msgLen)
+	rn, err = stream.Read(msgBytes[:])
+	if rn != int(msgLen) || err != nil {
+		err0 := fmt.Errorf("error reading response message from peer %s: %w", p2pPeer, err)
+		return peerMetaInfo{}, err0
+	}
+	var responseHeaders peerMetaHeaders
+	_, err = responseHeaders.UnmarshalMsg(msgBytes[:])
+	if err != nil {
+		err0 := fmt.Errorf("error unmarshaling response message from peer %s: %w", p2pPeer, err)
+		return peerMetaInfo{}, err0
+	}
+	headers := peerMetaHeadersToHTTPHeaders(responseHeaders)
+	matchingVersion, _ := checkProtocolVersionMatch(headers, netProtoSupportedVersions)
+	if matchingVersion == "" {
+		err0 := fmt.Errorf("peer %s does not support any of the supported protocol versions: %v", p2pPeer, netProtoSupportedVersions)
+		return peerMetaInfo{}, err0
+	}
+	return peerMetaInfo{
+		telemetryID:  headers.Get(TelemetryIDHeader),
+		instanceName: headers.Get(InstanceNameHeader),
+		version:      matchingVersion,
+		features:     headers.Get(PeerFeaturesHeader),
+	}, nil
+}
+
+func writePeerMetaHeaders(stream io.ReadWriter, p2pPeer peer.ID, networkProtoVersion string, pmp peerMetadataProvider) error {
+	header := make(http.Header)
+	setHeaders(header, networkProtoVersion, pmp)
+	meta := peerMetaHeadersFromHTTPHeaders(header)
+	data := meta.MarshalMsg(nil)
+	length := len(data)
+	if length > math.MaxUint16 {
+		// 64k is enough for everyone
```

Comment:
do we know this can NEVER be violated with a legit header?

Reply:
It is 250 bytes at the moment with all fields set. It gives quite a bit of room for extension and as well custom user values